core/src/parse.rs

```diff
@@ -96,177 +306,27 @@ where
-            ),
-            tag_with_args(
-                "CMP",
-                three_args(reg_ident, value_source, value_source),
-                |(dst, src_1, src_2)| Operator::Cmp(dst, src_1, src_2),
-            ),
-            tag_with_args(
-                "PUSH",
-                two_args(value_source, stack_ident),
-                |(src, stack)| Operator::Push(src, stack),
-            ),
-            tag_with_args(
-                "POP",
-                two_args(stack_ident, reg_ident),
-                |(stack, dst)| Operator::Pop(stack, dst),
-            ),
-        )),
-        Statement::Operator,
-    )(input)
-}
-
-fn jump_stmt(input: &str) -> ParseResult<'_, Statement> {
-    alt((
-        tag_with_args("JMP", one_arg(label), |l| Statement::Jump(Jump::Jmp, l)),
-        tag_with_args("JEZ", two_args(value_source, label), |(src, l)| {
-            Statement::Jump(Jump::Jez(src), l)
-        }),
-        tag_with_args("JNZ", two_args(value_source, label), |(src, l)| {
-            Statement::Jump(Jump::Jnz(src), l)
-        }),
-        tag_with_args("JGZ", two_args(value_source, label), |(src, l)| {
-            Statement::Jump(Jump::Jgz(src), l)
-        }),
-        tag_with_args("JLZ", two_args(value_source, label), |(src, l)| {
-            Statement::Jump(Jump::Jlz(src), l)
-        }),
-    ))(input)
-}
-
 /// Parses a line comment, which starts with a ; and runs to the end of the
 /// line. This terminates at the line ending, but does _not_ consume it.
-fn line_comment(input: &str) -> ParseResult<'_, &str> {
+fn line_comment(input: RawSpan) -> ParseResult<'_, ()> {
     map(
         context("Comment", preceded(char(';'), many0(is_not("\r\n")))),
-        |_| "", // throw the comment away
+        |_| (), // throw the comment away
     )(input)
 }
 
-/// Parses one source statement, not including surrounding whitespace.
-fn statement(input: &str) -> ParseResult<'_, Statement> {
-    context("Statement", alt((label_stmt, operator_stmt, jump_stmt)))(input)
-}
-
 /// Parse a single line, not including the line ending.
-fn line(input: &str) -> ParseResult<'_, Option<Statement>> {
-    context(
-        "Line",
-        delimited(space0, opt(statement), tuple((space0, opt(line_comment)))),
-    )(input)
-}
-
-/// Parse a full program
-fn parse_gdlk(input: &str) -> ParseResult<'_, Vec<Statement>> {
-    map(
-        // separated_list doesn't work properly so we have to do this
-        all_consuming(tuple((many0(terminated(line, line_ending)), opt(line)))),
-        // filter out None lines
-        |(lines, last_line)| {
-            lines
-                .into_iter()
-                .chain(iter::once(last_line.unwrap_or(None)))
-                .filter_map(std::convert::identity)
-                .collect()
-        },
+fn line(input: RawSpan) -> ParseResult<'_, Option<SpanNode<Statement<Span>>>> {
+    delimited(
+        space0,
+        opt(Statement::parse_node),
+        tuple((space0, opt(line_comment))),
```

Comment:
Im a little confused on how this handles the top level `alt` we had before. Does Nom locate just kinda handle that based on how the traits are implemented

Reply:
nvm just noticed the alt was in the trait impl for statement